examples/src/main/java/io/grpc/examples/errorhandling/DetailErrorSample.java

```diff
@@ -0,0 +1,243 @@
+import io.grpc.Metadata;
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import io.grpc.Status;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterBlockingStub;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterFutureStub;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterStub;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.protobuf.ProtoUtils;
+import io.grpc.stub.StreamObserver;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+/**
+ * Shows how to setting and reading RPC error details.
+ * Proto used here is just an example proto, but the pattern sending
+ * application error information as an application-specific binary protos
+ * in the response trailers is the recommended way to return application
+ * level error.
+ */
+public class DetailErrorSample {
+  static final Metadata.Key<DebugInfo>
+    DEBUG_INFO_TRAILER_KEY = ProtoUtils.keyForProto(DebugInfo.getDefaultInstance());
+
+  static final DebugInfo DEBUG_INFO =
+    DebugInfo.newBuilder()
+      .addStackEntries("stack_entry_1")
+      .addStackEntries("stack_entry_2")
+      .addStackEntries("stack_entry_3")
+      .setDetail("detailed error info.").build();
+
+  static final String DEBUG_DESC = "detailed error description";
+
+  public static void main(String[] args) throws Exception {
+    new DetailErrorSample().run();
+  }
+
+  private Server server;
+  private ManagedChannel channel;
+
+  void run() throws Exception {
+    server = ServerBuilder.forPort(0).addService(new GreeterGrpc.GreeterImplBase() {
+      @Override
+      public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
+        Metadata trailers = new Metadata();
+        trailers.put(DEBUG_INFO_TRAILER_KEY, DEBUG_INFO);
+        responseObserver.onError(Status.INTERNAL.withDescription(DEBUG_DESC)
+          .asRuntimeException(trailers));
+      }
+    }).build().start();
+    channel =
+      ManagedChannelBuilder.forAddress("localhost", server.getPort()).usePlaintext(true).build();
+
+    blockingCall();
+    futureCallDirect();
+    futureCallCallback();
+    asyncCall();
+    advancedAsyncCall();
+
+    channel.shutdown();
+    server.shutdown();
+    channel.awaitTermination(1, TimeUnit.SECONDS);
+    server.awaitTermination();
+  }
+
+  static void verifyErrorReply(Throwable t) {
+    Status status = Status.fromThrowable(t);
+    Metadata trailers = Status.trailersFromThrowable(t);
+    Verify.verify(status.getCode() == Status.Code.INTERNAL);
+    Verify.verify(trailers.containsKey(DEBUG_INFO_TRAILER_KEY));
+    Verify.verify(status.getDescription().equals(DEBUG_DESC));
+    try {
+      Verify.verify(trailers.get(DEBUG_INFO_TRAILER_KEY).equals(DEBUG_INFO));
+    } catch (IllegalArgumentException i) {
```

Comment:
From the `parseBytes` of `ProtoLiteUtils.metadataMarshaller` inside the `ProtoUtils.keyForProto`, maybe I should just drop it ?